host/fde_key_manager/fde_key_manager.c

```diff
@@ -0,0 +1,692 @@
+                                        strlen(FDE_JSON_RESULT_FEATURES));
+}
+
+int handle_fde_operation(char *request_str) {
+    struct json_object *request_json;
+    struct json_object *j_op;
+    const char *op;
+    json_bool j_ret;
+    int ret = EXIT_SUCCESS;
+
+    request_json = json_tokener_parse(request_str);
+    if (!request_json) {
+        ree_log(REE_ERROR, "Malformed json passed" );
+        return EXIT_FAILURE;
+    }
+    j_ret = json_object_object_get_ex(request_json,
+                                      FDE_JSON_OP,
+                                      &j_op);
+    if (j_ret == TRUE){
+        op = json_object_get_string(j_op);
+        // three operations are supported
+        // - FDE_JSON_OP_REVEAL
+        // - FDE_JSON_OP_LOCK
+        // - FDE_JSON_OP_INITIAL_SETUP
+        // - FDE_JSON_OP_FEATURES
+        ree_log(REE_INFO, "Handling operation: %s\n", op);
+        if (!strncmp(op, FDE_JSON_OP_REVEAL, strlen(FDE_JSON_OP_REVEAL))) {
+            ret = handle_operation_reveal(request_json);
+        } else if (!strncmp(op, FDE_JSON_OP_LOCK, strlen(FDE_JSON_OP_REVEAL))) {
+            ret = handle_operation_lock();
+        } else if (!strncmp(op, FDE_JSON_OP_INITIAL_SETUP, strlen(FDE_JSON_OP_INITIAL_SETUP))) {
+            ret = handle_operation_setup(request_json);
+        } else if (!strncmp(op, FDE_JSON_OP_FEATURES, strlen(FDE_JSON_OP_FEATURES))) {
+            ret = handle_operation_feature(request_json);
+        } else {
+            ree_log(REE_ERROR, "Unknown operation requested:[%s]", op );
+            ret = EXIT_FAILURE;
+        }
+    } else {
+        ree_log(REE_ERROR, "json: parse: op failed");
+    }
+    json_object_put(request_json);
+    return ret;
+}
+
+/**
+ * There are two fde hooks, 'fde-reveal-key' and 'fde-setup'
+ * 'fde-setup' hook can be called in 2 scenarios, as true snap hook
+ * and as hook within initranfs. Depending on the runtime environment, hook
+ * uses stdin/stdout or `snapctl` as way to retrieve input and pass the result.
+ * `snapctl fde-setup-request` and `snapctl fde-setup-result` are used in snap hook
+ * runtime environment.
+ * Input and output are always passed as formated json.
+ *
+ * - 'fde-reveal-key':
+ *   - supported operations:
+ *     - reveal:
+ *         - request: { "op": "reveal", "sealed-key": "base64-encoded-bytes",
+ *                 "handle": "base64-encoded-bytes"}
+ *         - result: {"key": "base64-encoded-bytes"}
+ *     - lock:
+ *         - request: { "op": "lock" }
+ *         - result:
+ *
+ * - 'fde-setup':
+ *   - supported operations:
+ *     - fde-setup:
+ *       - request: {"op": "initial-setup","key": "base64-encoded-bytes",
+ *              "key-name" : "string"}
+ *       - result: {"sealed-key": "base64-encoded-bytes",
+ *               "handle": "base64-encoded-bytes"}
+ *       - request: {"op":"features"}
+ *       - result: {"features": []}
+ */
+int main(int argc, char *argv[]) {
```

Comment:
Maybe it would be good to add a link to https://snapcraft.io/docs/uc20-fde-hooks in this comment, as it is a quite good explanation on how the program is used.

Reply:
Added